lbcluster/retry_module.go

```diff
@@ -0,0 +1,130 @@
+package lbcluster
+
+import (
+	"fmt"
+	"lb-experts/golbd/logger"
+	"time"
+)
+
+type Retry struct {
+	signal            chan int
+	done              chan bool
+	tick              chan bool
+	currentCount      int
+	maxCount          int
+	retryStarted      bool
+	maxDuration       time.Duration
+	retryDuration     time.Duration
+	prevRetryDuration time.Duration
+	logger            logger.Logger
+}
+
+const defaultMaxDuration = 5 * time.Minute
+
+func NewRetryModule(retryStartDuration time.Duration, logger logger.Logger) *Retry {
+	retry := &Retry{
+		maxCount:          -1,
+		currentCount:      0,
+		maxDuration:       defaultMaxDuration,
+		retryDuration:     retryStartDuration,
+		prevRetryDuration: retryStartDuration,
+		logger:            logger,
+	}
+	return retry
+}
+
+func (r *Retry) SetMaxDuration(maxDuration time.Duration) error {
+	if r.retryStarted {
+		return nil
+	}
+	if maxDuration <= 0 {
+		return fmt.Errorf("duration has to be greater than 0")
+	}
+	r.maxDuration = maxDuration
+	return nil
+}
+
+func (r *Retry) SetMaxCount(maxCount int) error {
+	if r.retryStarted {
+		return nil
+	}
+	if maxCount <= 0 {
+		return fmt.Errorf("max count has to be greater than 0")
+	}
+	r.maxCount = maxCount
+	return nil
+}
+
+func (r *Retry) start() {
+	r.retryStarted = true
+	signal := make(chan int)
+	done := make(chan bool)
+	end := time.Tick(r.maxDuration)
+	r.signal = signal
+	r.done = done
+	go func() {
+		for {
+			select {
+			case <-end:
+				r.done <- true
+				return
+			default:
+				if r.currentCount == r.maxCount {
+					r.done <- true
+					return
+				}
+			}
+		}
+	}()
+	r.run()
+}
+
+func (r *Retry) run() {
+	start := make(chan bool)
+	go func() {
```

Comment:
First of all, thanks for this contribution! I appreciate a lot that you took your time to look into the code, and think of ways of improving it. 

I'm concerned about the retry approach. If I get it right, the retry is done on a different goroutine. Imagine that this is going over 200 aliases, and, for some reason, there are issues with the DNS, and they start failing. All those retries will start piling up. Taking into account that the main loop retries every 10 seconds, this might turn into a DOS, isn't it?

I have the feeling that the current approach of failing once, and hoping that the next iteration will fix it might be less risky.

Reply:
Sure, the scenario does makes for the retry call to be removed. Should I remove the whole retry module or just the call (reserve the module for any future use maybe) ?

Reply:
I see. Yes I do have `computeNextRetryTime` method to compute for next retry time by using fibonacci algorithm. Would that suffice, or should I use another algorithm for that purpose?